### PR TITLE
[NEW] Add optional gloss coverage computation to lookup pipeline

### PR DESCRIPTION
**Pull request by GerrySant**
This PR isolates the functionality related to **gloss coverage computation** from the earlier, broader PR (#52).

As suggested in the discussion on the original PR
(see comment: [https://github.com/sign-language-processing/spoken-to-signed-translation/pull/52#issuecomment-3767395987](https://github.com/sign-language-processing/spoken-to-signed-translation/pull/52#issuecomment-3767395987)),
the changes have been split so that each PR focuses on a single concern.

#### What this PR includes

* Optional computation of **gloss coverage** (ratio of successfully matched glosses)
* Coverage is computed in `lookup_sequence()`
* Propagated through `gloss_to_pose()`
* Exposed via an opt-in CLI flag
* Default behavior and return types remain unchanged when the feature is disabled

#### What this PR intentionally does **not** include

* No changes to lookup heuristics or fallback strategies
* No verbosity / logging changes
* No threading or performance-related refactors
* No CLI behavior changes beyond the coverage flag

### CODE REVIEW

**Review by bricksdont — CHANGES_REQUESTED** [review]
thanks for this!

in my opinion modifying the output filename is impractical and confusing. The pose file path is a parser argument that a user supplies, and it would be confusing to deviate from that.

this computes coverage for each line of text separately, correct? which is the main reason why storing the stats is so awkward (because I'm assuming you didn't want to add a separate output file for the stats each time).

but the main application for this is computing the coverage over a large quantity of input lines. how about this:

a wrapper / separate CLI for the current line-by-line code that generates the poses in bulk. (I see there is some multiprocessing to parallelize lookups already, but an even better application of it would be processing separate sentences at the same time) not only could the generation be more efficient then, but also you could more easily add an argument:

```
--coverage-stats: path to a file where coverage statistics are saved
```

and then keeping track of the statistics can be implemented as a simple counter of an object:

```python
class Lookup:

   def __init__(self):
      self.total_tokens, self.matched_tokens = 0, 0

   def lookup():
      ...
      self.total_tokens += 1
      if match:
         self.matched_tokens += 1
```
(or implement it slightly differently, but the main point is that statistics could be accumulated somehow)

does this make sense?

minor:
- please extend the README as well
- increase the minor version for changes

File: spoken_to_signed/gloss_to_pose/lookup/fingerspelling_lookup.py [hunk 1]
```diff
@@ -44,15 +44,17 @@ def stretch_pose(self, pose: Pose, by: float) -> Pose:
...
+    def lookup(self, word: str, gloss: str, spoken_language: str, signed_language: str, source: str = None):
         if spoken_language not in self.words_index or signed_language not in self.words_index[spoken_language]:
             raise FileNotFoundError(
                 f"Language pair {spoken_language} -> {signed_language} not supported for fingerspelling"
             )
 
-        poses = list(self.characters_lookup(word.lower(), spoken_language, signed_language))
+        pairs = list(self.characters_lookup(word.lower(), spoken_language, signed_language))
+        keys = [key for key, _ in pairs]
+        poses = [pose for _, pose in pairs]
```
Comment (AmitMY) [inline spoken_to_signed/gloss_to_pose/lookup/fingerspelling_lookup.py]: use zip

**Review by AmitMY — CHANGES_REQUESTED** [review]
from my perspective - this PR includes multiple changes (at least 3).
please ask claude (or whichever ai tool) to extract the core of it to a new minimal PR.
and we can merge in chain. there's too many decisions made here that affect eachother

File: spoken_to_signed/gloss_to_pose/lookup/fingerspelling_lookup.py [hunk 2]
```diff
@@ -44,15 +44,17 @@ def stretch_pose(self, pose: Pose, by: float) -> Pose:
         pose.body.fps = fps
         return pose
 
-    def lookup(self, word: str, gloss: str, spoken_language: str, signed_language: str, source: str = None) -> Pose:
+    def lookup(self, word: str, gloss: str, spoken_language: str, signed_language: str, source: str = None):
```
Comment (bricksdont) [inline spoken_to_signed/gloss_to_pose/lookup/fingerspelling_lookup.py]: this function used to have a return value (`-> Pose`) could you add this back in?

**Review by bricksdont — APPROVED** [review]

File: spoken_to_signed/gloss_to_pose/lookup/lookup.py [hunk 3]
```diff
@@ -95,37 +98,67 @@ def lookup(self, word: str, gloss: str, spoken_language: str, signed_language: s
...
         # Backup strategy: revert to backup sign language
         if signed_language in LANGUAGE_BACKUP:
-            return self.lookup(word, gloss, spoken_language, LANGUAGE_BACKUP[signed_language], source)
+            pose, coverage_type, sub_elements = self.lookup(
+                word, gloss, spoken_language, LANGUAGE_BACKUP[signed_language], source
+            )
+            # If found in the backup language's own lexicon, label as language_backup; otherwise keep the type
+            if coverage_type == "lexicon":
+                return pose, "language_backup", None
+            return pose, coverage_type, sub_elements
```
Comment (AmitMY) [inline spoken_to_signed/gloss_to_pose/lookup/lookup.py]: use a named tuple?

**Review by AmitMY — CHANGES_REQUESTED** [review]

File: spoken_to_signed/gloss_to_pose/lookup/lookup.py [hunk 4]
```diff
@@ -95,37 +98,67 @@ def lookup(self, word: str, gloss: str, spoken_language: str, signed_language: s
...
+        spoken_language: str,
+        signed_language: str,
+        source: str = None,
+        coverage_info: bool = False,
+    ):
         def lookup_pair(pair):
             word, gloss = pair
             if word == "":
-                return None
+                return None, None, None, word, gloss
```
[the remark this answers is not in the document]
Comment (GerrySant) [reply spoken_to_signed/gloss_to_pose/lookup/lookup.py]: > unreadable. use a data class with default values?

Replaced with a `PairResult` dataclass with `pose`, `coverage_type`, and `sub_elements` defaulting to `None`.

**Review by AmitMY — CHANGES_REQUESTED** [review]
## PR Review — Gloss Coverage Computation

Overall this is a useful feature and the scope is now well-contained. Issues sorted by priority:

---

### 1. [CRITICAL] `matched` semantics are misleading and likely wrong

In `lookup.py:161`, `TokenCoverage` is created with `matched=(r.coverage_type == "lexicon")`. This means tokens resolved via `language_backup` or `fingerspelling_backup` are marked `matched=False`. But in `bin.py:_print_token_coverage`, these show as X even though they produced a valid pose. The coverage fraction only counts primary-lexicon hits, which is surprising and undocumented.

Either rename `matched` to `exact_lexicon_match`, or change it to mean "a pose was produced" (`matched=(r.coverage_type is not None)`).

---

### 2. [HIGH] `characters_lookup` can yield nothing causing crash

In `fingerspelling_lookup.py:53`, `keys, poses = zip(*self.characters_lookup(...))` will raise `ValueError` if the generator yields nothing (empty word, no matching characters). Add a guard for the empty case.

---

### 3. [HIGH] `text` passed to `CoverageStats.add_sentence` is the full input, not per-sentence

In `bin.py:171`, `args.text` is the entire input string. When there are multiple sentences, every JSON entry gets the same full text instead of the individual sentence text.

---

### 4. [HIGH] Coverage types should be an Enum, not magic strings

`coverage_type` values ("lexicon", "language_backup", "fingerspelling_backup", None) are scattered as raw strings across `lookup.py`, `fingerspelling_lookup.py`, `coverage.py`, and `visualize_coverage.py`. This is fragile — a typo in any one place silently breaks the logic. Define a `CoverageType` enum in `coverage.py` and use it everywhere. The color mapping and legend in `visualize_coverage.py` should also be driven by this enum rather than duplicating the strings.

---

### 5. [MEDIUM] Return type annotation is too vague

`gloss_to_pose/__init__.py:18` annotates return as `Union[Pose, tuple]`. Use `Union[Pose, tuple[Pose, list[TokenCoverage]]]`.

---

### 6. [MEDIUM] Boolean flag that changes return type is a code smell

The `coverage_info: bool` parameter changes the return type of multiple functions. Consider always returning a result object or having a separate method like `lookup_sequence_with_coverage()`.

---

### 7. [LOW] Lazy import inside `lookup_sequence`

In `lookup.py:155`, `TokenCoverage` is lazily imported inside the method body, but `coverage.py` is already imported at module level in `bin.py`. Since it has no heavy dependencies, import at the top of `lookup.py`.

---

### 8. [LOW] No unit tests for `CoverageStats` or `TokenCoverage`

The e2e tests are good, but there are no unit tests for `CoverageStats.fraction` (division-by-zero guard) or edge cases like empty gloss sequences.

---

### 9. [NIT] `None` as dict key in `_COLORS`

Works but is unconventional. With the enum suggestion above, this becomes a non-issue since you'd use e.g. `CoverageType.UNMATCHED` instead.

**Comment by AmitMY** [comment]
Thanks @GerrySant for the design and iteration on this — the feature is now in. Since `main` had moved underneath this branch (`lookup()` now returns `PoseResult`), a minimal re-implementation of the same functionality was rebuilt on top of it and merged as #78, crediting this PR: per-token coverage types, `--coverage-info` (with the color-coded terminal view built in) and `--coverage-stats` JSON output. Closing in favor of #78.
